src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,645 @@
+        /// This implementation simply explores the reachable space
+        /// by all sequences of `TestEvent`.
+        /// This function can be instanciated with `RetX` and `RetY`
+        /// among `NoRet` or `AllowRet` to resp. forbid/allow `x`/`y` to lose their
+        /// protector.
+        fn distinguishable<RetX, RetY>(&self, other: &Self) -> bool
+        where
+            RetX: Exhaustive + 'static,
+            RetY: Exhaustive + 'static,
+        {
+            if self == other { return false; }
+            let mut states = vec![(self.clone(), other.clone())];
+            let mut seen = FxHashSet::default();
+            while let Some(state) = states.pop() {
+                if !seen.insert(state.clone()) { continue; };
+                let (source, target) = state;
+                for evt in <TestEvent<RetX, RetY>>::exhaustive() {
+                    // Generate successor states through events (accesses and protector ends)
+                    let Ok(new_source) = source.clone().perform_test_event(&evt) else { continue; };
+                    let Ok(new_target) = target.clone().perform_test_event(&evt) else { return true; };
+                    if new_source == new_target { continue; }
+                    states.push((new_source, new_target));
+                }
+            }
+            false
+        }
+    }
+
+    #[test]
+    #[should_panic]
+    // This is why `Reserved -> Frozen` on foreign read for protected references
+    // prevents the insertion of spurious reads: the transition can cause UB in the target
+    // later down the line.
+    fn reserved_frozen_protected_distinguishable() {
+        let source = LocStateProtPair {
+            xy_rel: RelPosXY::MutuallyForeign,
+            x: LocStateProt {
+                state: LocationState::new(Permission::new_frozen()).with_access(),
+                prot: true,
+            },
+            y: LocStateProt {
+                state: LocationState::new(Permission::new_reserved(false)),
+                prot: true,
+            },
+        };
+        let acc = TestAccess { ptr: PtrSelector::X, kind: AccessKind::Read };
+        let target = source.clone().perform_test_access(&acc).unwrap();
+        assert!(source.y.state.permission.is_reserved(None));
+        assert!(target.y.state.permission.is_frozen());
+        assert!(!source.distinguishable::<(), ()>(&target))
+    }
+
+    #[derive(Clone, Debug)]
+    struct Pattern {
+        /// The relative position of `x` and `y` at the beginning of the pattern.
+        /// Might change if the opaque code contains any `retag y`.
+        xy_rel: RelPosXY,
+        /// Permission that `x` will be retagged as (always protected until
+        /// a possible `ret x` is the second phase).
+        x_retag_perm: LocationState,
+        /// Permission that `y` has at the beginning of the pattern.
+        /// Might by reset if the opaque code contains any `retag y`.
```

Comment:
```suggestion
        /// Might by reset if the opaque code contains any `retag y`.
        /// Can be any state, not necessarily initial
        /// (since `y` exists already before the pattern starts).
```